### ISSUE TITLE
error validating

### ISSUE THREAD

**Issue by woweenie** [bghira/SimpleTuner#546]
Training SD3 lora.

"[ERROR] (helpers.training.validation) Error generating validation image: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument mat1 in method wrapper_CUDA_addmm), Traceback (most recent call last):
  File "/workspace/train-sd3/SimpleTuner/helpers/training/validation.py", line 762, in validate_prompt
    validation_image_results = self.pipeline(**pipeline_kwargs).images"

If I put the VALIDATION_STEPS down to 10 then it works, but at the deafult 150 it just prints the error

**Comment by woweenie** [comment]
i think it happens when the VALIDATION_STEPS is longer than the number of steps int he epoch...

**Comment by OliviaOliveiira** [comment]
Had the very same error, and simple reboot fixed it for me

**Comment by mhirki** [comment]
I'm also seeing this issue but it's random. If I restart the script, the error may not show up at all. If I pass the option --validation_seed_source=gpu, I get a different error that is still random.

I ran the script with SIMPLETUNER_LOG_LEVEL=DEBUG and grabbed a debug.log where the error shows up. I also grabbed another debug.log where the error was not showing up. I didn't really see anything different between these myself though.

[debug-validation-error.zip](https://github.com/user-attachments/files/16045582/debug-validation-error.zip)
[debug-validation-success.zip](https://github.com/user-attachments/files/16045586/debug-validation-success.zip)

**Comment by bghira** [comment]
it does seem to be caused by having the generator on CPU:

```
2024-06-30 20:25:23,350 [DEBUG] (helpers.training.validation) Device for generator: cpu
2024-06-30 20:25:23,350 [DEBUG] (helpers.training.validation) Device for pooled_prompt_embeds: cuda:0
2024-06-30 20:25:23,350 [DEBUG] (helpers.training.validation) Device for negative_pooled_prompt_embeds: cuda:0
2024-06-30 20:25:23,350 [DEBUG] (helpers.training.validation) Device for prompt_embeds: cuda:0
2024-06-30 20:25:23,350 [DEBUG] (helpers.training.validation) Device for negative_prompt_embeds: cuda:0
2024-06-30 20:25:23,350 [DEBUG] (helpers.training.validation) Device for vae: cuda:0
2024-06-30 20:25:23,350 [DEBUG] (helpers.training.validation) Device for text_encoder_2: cpu
2024-06-30 20:25:23,350 [DEBUG] (helpers.training.validation) Device for text_encoder_3: cpu
2024-06-30 20:25:23,351 [DEBUG] (helpers.training.validation) Device for transformer: cuda:0
```

as the embeds and the model are all on cuda, the only way the input could end up on the cpu is because the generator is there? maybe the cpu generator needs to be moved to the device?

**Comment by mhirki** [comment]
I saw your commit https://github.com/bghira/SimpleTuner/commit/852e143dbcc6becfb422f85c310a24ab8e1de0d6 and tested it but looks like it's just crashing.

```
'torch._C.Generator' object has no attribute 'to'                                                                                                                                                                                                                                                               
Traceback (most recent call last):
  File "/nvme/home/mikaelh/Stable_Diffusion/bghira/SimpleTuner/train_sdxl.py", line 2255, in <module>
    main()
  File "/nvme/home/mikaelh/Stable_Diffusion/bghira/SimpleTuner/train_sdxl.py", line 2009, in main
    validation.run_validations(validation_type="intermediary", step=step)
  File "/nvme/home/mikaelh/Stable_Diffusion/bghira/SimpleTuner/helpers/training/validation.py", line 446, in run_validations
    self.process_prompts()
  File "/nvme/home/mikaelh/Stable_Diffusion/bghira/SimpleTuner/helpers/training/validation.py", line 654, in process_prompts
    self.validate_prompt(prompt, shortname, validation_input_image)
  File "/nvme/home/mikaelh/Stable_Diffusion/bghira/SimpleTuner/helpers/training/validation.py", line 687, in validate_prompt
    extra_validation_kwargs["generator"] = self._get_generator()
                                           ^^^^^^^^^^^^^^^^^^^^^
  File "/nvme/home/mikaelh/Stable_Diffusion/bghira/SimpleTuner/helpers/training/validation.py", line 180, in _get_generator
    return _generator.to(self.accelerator.device)
           ^^^^^^^^^^^^^
AttributeError: 'torch._C.Generator' object has no attribute 'to'
```

**Comment by bghira** [comment]
pytorch 🤦 reverted

**Comment by bghira** [comment]
if anyone is experiencing this reliably, please alter the local version of the pipeline to print the devices of each input on the offending line. i still can't recreate this.

**Comment by OliviaOliveiira** [comment]
I’m not receiving it any longer after fixing my batch size * gradient
accumulation steps + I found out that .jpeg doesn’t really work either,
though simply renaming it to jpg solves the issue


Чт, 4 июля 2024 г. в 4:29 PM, Bagheera ***@***.***>:

**Comment by OliviaOliveiira** [comment]
> if anyone is experiencing this reliably, please alter the local version of the pipeline to print the devices of each input on the offending line. i still can't recreate this.

Can yoг elaborate more simply?(e.g. execute this or that command?) 
I noticed that I get this error after certain failures (like forgot to change the dataset folder, or errors in huggingface push), and then when the training runs succesfully, validation gets this error

**Comment by bghira** [comment]
it will mention in the traceback that there is a method where it fails on in pipeline.py which you should open up and add something like

```py
print(f"The device for input is {input.device}")
```

for each input there

**Comment by bghira** [comment]
this should be fixed by #562
